auth/jwt-token-manager-v1.ts

```diff
@@ -0,0 +1,202 @@
+    // all parameters are optional
+    options = options || {} as Options;
+
+    this.tokenInfo = {};
+    this.tokenName = 'access_token';
+
+    if (options.url) {
+      this.url = options.url;
+    }
+
+    if (options.accessToken) {
+      this.userAccessToken = options.accessToken;
+    }
+
+    this.requestWrapperInstance = new RequestWrapper();
+  }
+
+  /**
+   * This function sends an access token back through a callback. The source of the token
+   * is determined by the following logic:
+   * 1. If user provides their own managed access token, assume it is valid and send it
+   * 2. a) If this class is managing tokens and does not yet have one, make a request for one
+   *    b) If this class is managing tokens and the token has expired, request a new one
+   * 3. If this class is managing tokens and has a valid token stored, send it
+   *
+   * @param {Function} cb - callback function that the token will be passed to
+   */
+  public getToken(cb: Function) {
+    if (this.userAccessToken) {
+      // 1. use user-managed token
+      return cb(null, this.userAccessToken);
+    } else if (!this.tokenInfo[this.tokenName] || this.isTokenExpired()) {
+      // 2. request a new token
+      this.requestToken((err, tokenResponse) => {
+        if (!err) {
+          try {
+            this.saveTokenInfo(tokenResponse);
+          } catch(e) {
+            // send lower level error through callback for user to handle
+            err = e;
+          }
+        }
+        // return null for access_token if there is an error
+        return cb(err, this.tokenInfo[this.tokenName] || null);
+      });
+    } else {
+      // 3. use valid, sdk-managed token
+      return cb(null, this.tokenInfo[this.tokenName]);
+    }
+  }
+
+  /**
+   * Set a self-managed access token.
+   * The access token should be valid and not yet expired.
+   *
+   * By using this method, you accept responsibility for managing the
+   * access token yourself. You must set a new access token before this
+   * one expires. Failing to do so will result in authentication errors
+   * after this token expires.
+   *
+   * @param {string} accessToken - A valid, non-expired access token
+   * @returns {void}
+   */
+  public setAccessToken(accessToken: string): void {
+    this.userAccessToken = accessToken;
+  }
+
+  /**
+   * Request a JWT using an API key.
+   *
+   * @param {Function} cb - The callback that handles the response.
+   * @returns {void}
+   */
+  protected requestToken(cb: Function): void {
+    const err = new Error('`requestToken` MUST be overridden by a subclass of JwtTokenManagerV1.');
+    cb(err, null);
+  }
+
+  /**
+   * Compute and return a Basic Authorization header from a username and password.
+   *
+   * @param {string} username - The username or client id
+   * @param {string} password - The password or client secret
+   * @returns {string}
+   */
+  protected computeBasicAuthHeader(username, password): string {
```

Comment:
I still think this method is out of place in this file.  Maybe it belongs in it's own file: `auth/basic-v1.ts`?

Reply:
That's a good idea. This logic is used in the base class too, so putting it in a utility type file and sharing it makes sense.